### ISSUE TITLE
Real time inference problems

### ISSUE THREAD

**Issue by Wolfhan99** [aask1357/hilcodec#6]
> Regarding training,
> HILCodec is composed of a small encoder and a large decoder. I suggest to reduce the `channels_dec` to 32 and use higher bitrate. (If the mixed-precision training fails, you may try to reduce the learning rate.)
> Also, our colleagues found that applying FINITE SCALAR QUANTIZATION to HiFi-Codec improves the speech quality at ultra-low bitrates. It would be non-trivial, but I think it is worth trying to replace the RVQ to FSQ for HILCodec. (https://arxiv.org/pdf/2309.15505)
> 
> Regarding inference,
> By default, HILCodec encodes and decodes 320 samples at a time @ 24kHz. This introduces a delay of 13.3ms + computation time. If increasing the delay is acceptable, changing the codec to encode and decode more samples at once will speed up the RTF. (The number of samples processed at a time must be a multiple of 320, such as 640, 960, etc. If you want to process a different number of samples at once, you can adjust the strides of [2, 4, 5, 8] to other values and then train. Currently, 1 frame corresponds to 2x4x5x8=320 samples.) 

 _Originally posted by @aask1357 in [#5](https://github.com/aask1357/hilcodec/issues/5#issuecomment-2644559945)_
Hello @aask1357,

First of all, thank you for your great work!

I have a question regarding real-time inference using hilcodec. As you mentioned, the number of samples processed at a time must be a multiple of the model's downsampling factor.

In my scenario, the audio sampling rate is 8kHz, and the current downsampling factor of the model is 320. This means that the processing time for one time step is 40ms (since 320 samples at 8kHz correspond to 40ms).

However, I would like to reduce the processing time to 20ms. Does this mean I need to adjust the model's downsampling factor to 160?

Alternatively, I tried another approach: setting the frame shift for audio processing during real-time inference to 20ms (160 samples), while keeping the network's input size at 40ms (320 samples). However, the reconstructed audio quality using this method is very poor, with numerous artifacts. Here is the code I used (not thoroughly checked):

```python

hop_size = 320
n_quantizer = 4
time_shift = hop_size // 2

for start in range(0, length, frame_shift):
    end = start + hop_size
    if end > length:
        wav_in = wav[:, :, start:length]
        wav_in = F.pad(wav_in, (0, hop_size - (end - length)))
    else:
        wav_in = wav[:, :, start:end]
    x, cache_enc = model.encoder(wav_in, cache_enc)
    indices = model.quantizer(x, n=n_quantizer) 
    indices_list.append(indices)

for indices in indices_list:
    q = model.dequantizer(indices, n_quantizer)
    wav_out, cache_dec = model.decoder(q, *cache_dec)
    wav_out_list.append(wav_out.squeeze(1))

for index, wav in enumerate(wav_out_list):
    if index == 0:
        wav_out_list2.append(wav)
    else:
        wav_out_list2.append(wav[..., -frame_shift:]) # add the last frame_shift samples
wav_out = torch.cat(wav_out_list2, dim=1)
    
# then write the wav_out tensor to audio. The audio quality is low and contains many artifacts.
```

Could you please provide some guidance on how to achieve better audio quality while maintaining a 20ms processing time? Any suggestions would be greatly appreciated!

**Comment by aask1357** [comment]
I recommend adjusting the downsampling factor to 160 and training the model again.  With a model trained using a downsampling ratio of 320, changing the frame shift during inference will definitely result in strange output.   
Additionally, using a model trained at 24kHz (our default checkpoint) to process 8kHz audio will also lead to strange output.

**Comment by Wolfhan99** [comment]
Thanks for your reply.

I've retrained the model base on 8kHz audio and adjusted the dwonsampling factor to 160. 
Comparing to 320 downsampling factor, the model performance degraed a lot particularly in low biterate(in this case, using 2 codebook, quantizer bit=8, biterate= 8000/160\*8\*2=800bps).  **The pesq score dropped about 0.4**.

Due to certain constraints, **the bitrate is limited to 800 bps and the processing time is capped at 20ms** in some scenarios. Do you have any suggestions on how improve this situation?

**Comment by aask1357** [comment]
I think the objective scores, including PESQ, cannot fully approximate the subjective quality. In my early experiments, using a higher weight for reconstruction loss and lower weights for GAN losses resulted in higher objective scores, but lower subjective quality. Also, wideband PESQ and narrowband PESQ are different, which can be the reason for the decreased PESQ score (I'm not sure about this). So, if you want a higher PESQ, you can increase the weight for the reconstruction loss. However, I guess that your concern is the subjective quality. I assume you've already listened to the output samples, and perhaps the sound quality was poor.

I think 800bps is too harsh. If your computational budget is enough, I think the best way (and maybe the only way) to increase the audio quality at ultra-low bitrate is to increase the model size. The easiest way is to increase the `channels_enc` and `channels_dec`. You can also increase the `n_residual_enc` and `n_residual_dec`. In this case, I recommend setting `res_scale_enc=rsqrt(n_residual_enc)` and `res_scale_dec=rsqrt(n_residual_dec)`. OR you can try other open-source codecs. If you don't need streaming codec, you can try wavtokenizer (https://github.com/jishengpeng/WavTokenizer).

If your computational budget is not enough, I think it would be really hard to increase the quality at 800bps.
